src/libsyntax/ptr.rs

```diff
@@ -101,6 +101,7 @@ impl<T: 'static> P<T> {
                 // Recreate self from the raw pointer.
                 Some(P { ptr: Box::from_raw(p) })
             } else {
+                drop(Box::from_raw(p));
```

Comment:
This is wrong, because of the `ptr::read(p)` call above - you need to deallocate *without* dropping the `T` value inside the `Box`.